paperqa/agents/__init__.py

```
@@ -0,0 +1,500 @@
+            # If both dict_a and dict_b have a dict for this key, recurse
+            result[key] = merge_dicts(result[key], value)
+        else:
+            # Otherwise, just update the value
+            result[key] = value
+
+    return result
+
+
+def get_merged_settings(
+    settings: dict[str, Any], settings_path: Path | None = None
+) -> dict[str, Any]:
+    """Merges a new settings with the current settings saved to file."""
+    if settings_path is None:
+        settings_path = pqa_directory("settings") / "settings.yaml"
+
+    current_settings = get_current_settings(settings_path)
+
+    # deal with the nested key case
+    return merge_dicts(current_settings, settings)
+
+
+@app.command("set")
+def set_setting(
+    variable: Annotated[
+        str,
+        typer.Argument(
+            help=(
+                "PaperQA variable to set, see agents.models.QueryRequest object for all settings, "
+                "nested options can be set using periods, ex. agent_tools.paper_directory"
+            )
+        ),
+    ],
+    value: Annotated[
+        str,
+        typer.Argument(
+            help=(
+                "Value to set to the variable, will be cast to the correct type automatically."
+            )
+        ),
+    ],
+) -> bool:
+    """Set a persistent PaperQA setting."""
+    settings_path = pqa_directory("settings") / "settings.yaml"
+
+    current_settings = get_merged_settings(
+        parse_dot_to_dict(variable, value), settings_path=settings_path
+    )
+
+    try:
+        _ = QueryRequest(**current_settings)
+    except MismatchedModelsError:
+        pass
+    except ValueError as e:
+        raise ValueError(
+            f"{variable} (with value {value}) is not a valid setting."
+        ) from e
+
+    logger.info(f"{variable} set to {str(value)[:100]}!")
+
+    with open(settings_path, "w") as f:
+        yaml.dump(current_settings, f)
+
+    return True
```

Let's not return anything, in validators it's best to:
- Fail and blow up with an `Exception`
- Silently succeed on success